### PR TITLE
federation: use generated listers

### PR DESCRIPTION
**Pull request by ncdc**
**What this PR does / why we need it**: switch federation code to use generated listers

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

cc @smarterclayton @derekwaynecarr @marun @gmarek @wojtek-t @deads2k @liggitt @sttts @timothysc @nikhiljindal @jianhuiz @kubernetes/sig-federation-pr-reviews @kubernetes/sig-scalability-pr-reviews

### CODE REVIEW

File: federation/pkg/federation-controller/service/endpoint_helper.go [hunk 1]
```diff
@@ -81,29 +81,24 @@ func (cc *clusterClientCache) syncEndpoint(key, clusterName string, clusterCache
...
-	} else {
+	endpoint, err := clusterCache.endpointStore.Endpoints(namespace).Get(name)
+	switch {
+	case errors.IsNotFound(err):
 		// service absence in store means watcher caught the deletion, ensure LB info is cleaned
 		glog.Infof("Can not get endpoint %v for cluster %s from endpointStore", key, clusterName)
 		err = cc.processEndpointDeletion(cachedService, clusterName, serviceController)
+	case err != nil:
+		glog.Errorf("Did not successfully get %v from store: %v, will retry later", key, err)
+		clusterCache.endpointQueue.Add(key)
```
Comment (deads2k) [inline federation/pkg/federation-controller/service/endpoint_helper.go]: this case used to return

File: federation/pkg/federation-controller/service/endpoint_helper.go [hunk 2]
```diff
@@ -81,29 +81,24 @@ func (cc *clusterClientCache) syncEndpoint(key, clusterName string, clusterCache
...
 		clusterCache.endpointQueue.Add(key)
 		return err
 	}
-	if exists {
-		endpoint, ok := endpointInterface.(*v1.Endpoints)
-		if ok {
-			glog.V(4).Infof("Found endpoint for federation service %s/%s from cluster %s", endpoint.Namespace, endpoint.Name, clusterName)
-			err = cc.processEndpointUpdate(cachedService, endpoint, clusterName, serviceController)
-		} else {
-			_, ok := endpointInterface.(cache.DeletedFinalStateUnknown)
```
Comment (deads2k) [inline federation/pkg/federation-controller/service/endpoint_helper.go]: This could never actually have happened before, rigth?

**Comment by deads2k** [comment]
a slipped case that would fall through correctly I think, one question.  lgtm otherwise.

File: federation/pkg/federation-controller/service/endpoint_helper.go (hunk 2 above)
Comment (ncdc) [inline federation/pkg/federation-controller/service/endpoint_helper.go]: I honestly don't remember in what situation(s) a tombstone is possible

File: federation/pkg/federation-controller/service/endpoint_helper.go (hunk 1 above)
Comment (ncdc) [inline federation/pkg/federation-controller/service/endpoint_helper.go]: @deads2k fixed, ptal

**Comment by deads2k** [comment]
lgtm

File: federation/pkg/federation-controller/service/endpoint_helper.go [hunk 3]
```diff
@@ -81,29 +81,25 @@ func (cc *clusterClientCache) syncEndpoint(key, clusterName string, clusterCache
...
+	endpoint, err := clusterCache.endpointStore.Endpoints(namespace).Get(name)
+	switch {
+	case errors.IsNotFound(err):
 		// service absence in store means watcher caught the deletion, ensure LB info is cleaned
 		glog.Infof("Can not get endpoint %v for cluster %s from endpointStore", key, clusterName)
 		err = cc.processEndpointDeletion(cachedService, clusterName, serviceController)
+	case err != nil:
+		glog.Errorf("Did not successfully get %v from store: %v, will retry later", key, err)
+		clusterCache.endpointQueue.Add(key)
+		return err
```
Comment (marun) [inline federation/pkg/federation-controller/service/endpoint_helper.go]: (no action required) I don't think returning an error from this function makes any sense. The only thing the caller does is log it at Info level, and it's already logged at Error level.

File: federation/pkg/federation-controller/replicaset/replicasetcontroller.go [hunk 4]
```diff
@@ -349,12 +353,19 @@ func (frsc *ReplicaSetController) deliverLocalReplicaSet(obj interface{}, durati
...
 	if err != nil {
+		glog.Errorf("Error splitting key for object %v: %v", obj, err)
+	}
+	_, err = frsc.replicaSetLister.ReplicaSets(namespace).Get(name)
+	switch {
+	case errors.IsNotFound(err):
+		// do nothing
+	case err != nil:
 		glog.Errorf("Couldn't get federation replicaset %v: %v", key, err)
 		return
```
Comment (marun) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller.go]: This return statement seems unnecessary.

Comment (ncdc) [inline federation/pkg/federation-controller/replicaset/replicasetcontroller.go]: Agreed; it's a holdover from the previous code that wasn't using a switch.

**Comment by nikhiljindal** [comment]
I havent looked at the code closely, but the change looks good at a high level.
/approve

I will let the existing reviewers do a detailed code review and let them LGTM.

**Comment by ncdc** [comment]
@marun I pushed an update to get rid of that superfluous `return`

**Comment by marun** [comment]
lgtm

**Comment by ncdc** [comment]
@nikhiljindal @marun who else (if anyone) should review this?

**Comment by marun** [comment]
@madhusudancs Please review

**Comment by ncdc** [comment]
@k8s-bot kops aws e2e test this

**Comment by ncdc** [comment]
Any other comments? Can we lgtm this?

File: federation/pkg/federation-controller/service/endpoint_helper.go [hunk 5]
```diff
@@ -81,29 +81,25 @@ func (cc *clusterClientCache) syncEndpoint(key, clusterName string, clusterCache
...
-		}
-	} else {
+	endpoint, err := clusterCache.endpointStore.Endpoints(namespace).Get(name)
+	switch {
+	case errors.IsNotFound(err):
 		// service absence in store means watcher caught the deletion, ensure LB info is cleaned
 		glog.Infof("Can not get endpoint %v for cluster %s from endpointStore", key, clusterName)
 		err = cc.processEndpointDeletion(cachedService, clusterName, serviceController)
+	case err != nil:
+		glog.Errorf("Did not successfully get %v from store: %v, will retry later", key, err)
```
Comment (smarterclayton) [inline federation/pkg/federation-controller/service/endpoint_helper.go]: util.HandleError.

File: federation/pkg/federation-controller/service/service_helper.go [hunk 6]
```diff
@@ -85,31 +84,26 @@ func (cc *clusterClientCache) syncService(key, clusterName string, clusterCache
...
-		}
-	} else {
+	service, err := clusterCache.serviceStore.Services(namespace).Get(name)
+	switch {
+	case errors.IsNotFound(err):
 		glog.Infof("Can not get service %v for cluster %s from serviceStore", key, clusterName)
 		needUpdate = cc.processServiceDeletion(cachedService, clusterName)
 		isDeletion = true
+	case err != nil:
+		glog.Errorf("Did not successfully get %v from store: %v, will retry later", key, err)
```
Comment (smarterclayton) [inline federation/pkg/federation-controller/service/service_helper.go]: util.HandleError

File: federation/pkg/federation-controller/service/servicecontroller.go [hunk 7]
```diff
@@ -938,44 +940,40 @@ func (s *ServiceController) syncService(key string) error {
...
 		glog.Infof("Service has been deleted %v", key)
 		err, retryDelay = s.processServiceDeletion(key)
-	}
-	// Create a copy before modifying the obj to prevent race condition with
-	// other readers of obj from store.
-	obj, err := conversion.NewCloner().DeepCopy(objFromStore)
-	if err != nil {
-		glog.Errorf("Error in deep copying service %v retrieved from store: %v", key, err)
+	case err != nil:
+		glog.Errorf("Unable to retrieve service %v from store: %v", key, err)
```
Comment (smarterclayton) [inline federation/pkg/federation-controller/service/servicecontroller.go]: handleError everywhere

File: federation/pkg/federation-controller/service/endpoint_helper.go (hunk 5 above)
Comment (ncdc) [inline federation/pkg/federation-controller/service/endpoint_helper.go]: That would require sweeping through the entire package and doing a find/replace. I opted not to do that here and would prefer to do it in a follow up if that's ok?

Comment (smarterclayton) [inline federation/pkg/federation-controller/service/endpoint_helper.go]: You added new code?

**Comment by smarterclayton** [comment]
I'm fine with a follow up, but I do expect glog.Errorf to stop being added even in minor stuff.  Needs to be linted for, really.

**Comment by ncdc** [comment]
I'll find or create an issue, and do a sweep of the repo

File: federation/pkg/federation-controller/service/endpoint_helper.go (hunk 5 above)
Comment (ncdc) [inline federation/pkg/federation-controller/service/endpoint_helper.go]: "ish". Before, it just did `clusterCache.endpointStore.GetByKey` and if that errored, it did the `glog.Errorf` in question. Now, it's a 2-parter: `cache.SplitMetaNamespaceKey` followed by `clusterCache.endpointStore.Endpoints(namespace).Get(name)`, both of which could error.

**Pull request merged**